agent/consul/stream/event_publisher.go

```diff
@@ -0,0 +1,305 @@
+}
+
+// NewEventPublisher returns an EventPublisher for publishing change events.
+// Handlers are used to convert the memDB changes into events.
+// A goroutine is run in the background to publish events to all subscribes.
+// Cancelling the context will shutdown the goroutine, to free resources,
+// and stop all publishing.
+func NewEventPublisher(ctx context.Context, handlers SnapshotHandlers, snapCacheTTL time.Duration) *EventPublisher {
+	e := &EventPublisher{
+		snapCacheTTL: snapCacheTTL,
+		topicBuffers: make(map[Topic]*eventBuffer),
+		snapCache:    make(map[Topic]map[string]*eventSnapshot),
+		publishCh:    make(chan changeEvents, 64),
+		subscriptions: &subscriptions{
+			byToken: make(map[string]map[*SubscribeRequest]*Subscription),
+		},
+		snapshotHandlers: handlers,
+	}
+
+	go e.handleUpdates(ctx)
+
+	return e
+}
+
+// PublishEvents to all subscribers. tx is a read-only transaction that captures
+// the state at the time the change happened. The caller must never use the tx once
+// it has been passed to PublishChanged.
+func (e *EventPublisher) PublishEvents(events []Event) {
+	if len(events) > 0 {
+		e.publishCh <- changeEvents{events: events}
+	}
+}
+
+func (e *EventPublisher) handleUpdates(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			// TODO: also close all subscriptions so the subscribers are moved
+			// to the new publisher?
+			return
+		case update := <-e.publishCh:
+			e.sendEvents(update)
+		}
+	}
+}
+
+// sendEvents sends the given events to any applicable topic listeners, as well
+// as any ACL update events to cause affected listeners to reset their stream.
+func (e *EventPublisher) sendEvents(update changeEvents) {
+	for _, event := range update.events {
+		if unsubEvent, ok := event.Payload.(closeSubscriptionPayload); ok {
+			e.subscriptions.closeSubscriptionsForTokens(unsubEvent.tokensSecretIDs)
+		}
+	}
+
+	eventsByTopic := make(map[Topic][]Event)
+	for _, event := range update.events {
+		eventsByTopic[event.Topic] = append(eventsByTopic[event.Topic], event)
+	}
+
+	e.lock.Lock()
+	defer e.lock.Unlock()
+	for topic, events := range eventsByTopic {
+		e.getTopicBuffer(topic).Append(events)
+	}
+}
+
+// getTopicBuffer for the topic. Creates a new event buffer if one does not
+// already exist.
+//
+// EventPublisher.lock must be held to call this method.
+func (e *EventPublisher) getTopicBuffer(topic Topic) *eventBuffer {
+	buf, ok := e.topicBuffers[topic]
+	if !ok {
+		buf = newEventBuffer()
+		e.topicBuffers[topic] = buf
+	}
+	return buf
+}
+
+// Subscribe returns a new stream.Subscription for the given request. A
+// subscription will stream an initial snapshot of events matching the request
+// if required and then block until new events that modify the request occur, or
+// the context is cancelled. Subscriptions may be forced to reset if the server
+// decides it can no longer maintain correct operation for example if ACL
+// policies changed or the state store was restored.
```

Comment:
> or the state store was restored

Huh interesting yeah I did consider it then but possibly didn't actually code that part?

Reply:
Previously there was no way to close down the EventPublisher goroutine, and I don't think subscriptions were closed